Sources/CombineX/Internal/Cache.swift

```diff
@@ -0,0 +1,74 @@
+import CXUtility
+
+class Cache<Key: Hashable, Value> {
+    
+    private var storage: [Key: Value] = [:]
+    
+    private var lock = Lock()
+    
+    func value(for key: Key) -> Value? {
+        return storage[key]
+    }
+    
+    func value(for key: Key, make: () throws -> Value) rethrows -> Value {
+        if let value = storage[key] {
+            return value
+        }
+        lock.lock()
+        defer { lock.unlock() }
+        if let value = storage[key] {
+            // avoid double write
+            return value
+        }
+        let value = try make()
+        storage[key] = value
+        return value
+    }
```

Comment:
I think we could use `if let value = storage[key]` without locking to return early ONLY IF:

- Once the value is added, it will not be removed. (So we have a high probability of returning early.)
- The same key always corresponds to the same value.

Otherwise, this trick won't bring much performance improvement.

Reply:
So, if this class is an ad-hoc class, for example, `TypeInfoCache`, everything is understandable. : )